### PR TITLE
Some fixes and add support for importing from nested properties and block scoped imports

### PR DESCRIPTION
**Pull request by dhharker**
(Ready for review)

This addresses a few issues I ran into while trying to use `js-to-styles-var-loader` where it didn't quite behave as I had expected.

 - [x] Fix issue where build would fail if last 'require' in a file was inside a block scope
 - [x] Fix issue where 'require's inside a block scope were included outside of that scope
 - [x] Add ability to access arbitrarily deep properties within a require()'d file e.g. `require("./some-vars.js").deeply.nested.properties`
 - [x] (Update tests)

### CODE REVIEW

**Comment by tompascall** [comment]
Hi! Thanks for your improvements, I'll review them soon

File: __tests__/index.spec.js [hunk 1]
```diff
@@ -53,42 +53,22 @@ describe('js-to-styles-vars-loader', () => {
...
-            expect(operator.getModulePath('require("./mocks/colors.js");\n')).toEqual([{path: "./mocks/colors.js"}]);
-
-            expect(operator.getModulePath('require("./mocks/colors.js");\n' + 'require("./mocks/sizes.js");')).toEqual([{path: "./mocks/colors.js"}, {path:"./mocks/sizes.js"}]);
-
-            expect(operator.getModulePath('require("./mocks/corners.js").typeTwo;\n')).toEqual([{path: "./mocks/corners.js", methodName: 'typeTwo'}]);
-        });
+    describe('guardExportType', () => {
+      it ("throws on anything except an object, does not throw otherwise", () => {
+        const areOk = [{}, {a: "foo"}];
+        const areNotOk = [[], ["a"], "", "123", 123, false, true];
```
Comment (tompascall) [inline __tests__/index.spec.js]: I suggest also testing null, undefined, NaN

File: __tests__/index.spec.js [hunk 2]
```diff
@@ -98,26 +78,63 @@ describe('js-to-styles-vars-loader', () => {
...
-        });
-
-        it('handles methodName if it is given', () => {
-            const varData = operator.getVarData([{ path:'./mocks/corners.js', methodName: 'typeOne'}], context);
+        it('uses value from property', () => {
+            const varData = operator.getVarData(path.join(context.context, './mocks/corners.js'), 'typeOne');
             expect(varData).toEqual({ tiny: '1%', medium: '3%'});
         });
+        it('uses value from nested property', () => {
+            const varData = operator.getVarData(path.join(context.context, './mocks/corners.js'), 'deep.nested');
```
Comment (tompascall) [inline __tests__/index.spec.js]: I suggest using either using a fake module that uses such a require or test the property getting part of regex (because this case is not covered yet)

File: __tests__/index.spec.js [hunk 3]
```diff
@@ -139,22 +156,39 @@ describe('js-to-styles-vars-loader', () => {
             context: path.resolve(),
             addDependency () {}
         };
+        const content = "require('./mocks/colors.js');\n" +
+            ".someClass { color: #fff;}";
+
+        const trimmer = (str) => {
+          return (str.split("\n").filter(a => a).map(s => s.trim()).join(" ")).trim()
```
Comment (tompascall) [inline __tests__/index.spec.js]: That filter with identity function seems unnecessary.

File: index.js [hunk 4]
```diff
@@ -1,48 +1,30 @@
 const path = require('path');
 const decache = require('decache');
+const squba = require('squba')
 
-const requireReg = /require\s*\(['|"](.+)['|"]\)(?:\.([^;\s]+))?[;\s]/g;
+const requireReg = /require\s*\((["'])([\w.\/]+)(?:\1)\)((?:\.[\w_-]+)*);?/igm;
```
Comment (tompascall) [inline index.js:5]: It is really cool that we can get nested properties. The thing is that I tried it in our project, and it failed, because the property got by the new regex started with a dot (like '.default'). I tried it with this modified regex and worked: ```/require\s*\((["'])([\w.\/]+)(?:\1)\)\.((?:[\w_-]+)*);?/igm``` (I moved the ```\.``` out of the group)

File: index.js [hunk 5]
```diff
@@ -1,48 +1,30 @@
...
-        const reg = new RegExp(requireReg);
-        const modulePaths = [];
-        let match;
-        while (match = reg.exec(modulePart)) {
-            modulePaths.push({
-                path: match[1],
-                methodName: match[2]
-            });
+    guardExportType (data, relativePath) {
+        if (typeof data !== "object" || Array.isArray(data)) {
```
Comment (tompascall) [inline index.js]: If we'll check null it must be updated, maybe like this: ```if ({}.toString.call(data) == '[object Object]')```

**Review by tompascall — CHANGES_REQUESTED** [review]

File: index.js (hunk 4 above)
Comment (dhharker) [reply index.js:5]: Strange, it is working for me I think (although I will double check now). At a glance that modified regex looks like it will not match a `require("...");` with no dot-part. I'll sort it out one way or another, anyway, I agree it is a bit nasty passing the key string all the way through to `squba` with the `.` still stuck on the front.

File: __tests__/index.spec.js (hunk 3 above)
Comment (dhharker) [reply __tests__/index.spec.js]: Yes I think it is, has been left in by accident.

**Comment by dhharker** [comment]
Thanks for the quick review @tompascall. I will try and address these issues in the next few days when I have a bit of time.

**Comment by tompascall** [comment]
Ok, thank you!

On 16 Jan 2018 11:59 p.m., "David Harker" <notifications@github.com> wrote:

Thanks for the quick review @tompascall <https://github.com/tompascall>. I
will try and address these issues in the next few days when I have a bit of
time.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/tompascall/js-to-styles-var-loader/pull/6#issuecomment-358136365>,
or mute the thread

File: __tests__/index.spec.js (hunk 1 above)
Comment (dhharker) [reply __tests__/index.spec.js]: :+1:

File: __tests__/index.spec.js (hunk 2 above)
Comment (dhharker) [reply __tests__/index.spec.js]: :+1:

File: index.js (hunk 4 above)
Comment (dhharker) [reply index.js:5]: :+1:

**Review by tompascall — APPROVED** [review]
